### PR TITLE
Action freshdesk

### PR DESCRIPTION
**Pull request by bush4u**

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/freshdesk/actions/create-company/create-company.mjs [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import { removeNullEntries } from "../../common/utils.mjs";
+import freshdesk from "../../freshdesk.app.mjs";
+
+export default {
+  key: "freshdesk-create-company",
+  name: "Create a Company",
+  description: "Create a company. [See docs here](https://developers.freshdesk.com/api/#companies)",
+  version: "0.0.2",
```
Comment (lcaresia) [inline components/freshdesk/actions/create-company/create-company.mjs:8]:
```suggestion
  version: "0.0.1",
```

File: components/freshdesk/actions/create-contact/create-contact.mjs [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import { removeNullEntries } from "../../common/utils.mjs";
+import freshdesk from "../../freshdesk.app.mjs";
+
+export default {
+  key: "freshdesk-create-contact",
+  name: "Create a Contact",
+  description: "create a contact. [See docs here](https://developers.freshdesk.com/api/#companies)",
```
Comment (lcaresia) [inline components/freshdesk/actions/create-contact/create-contact.mjs:7]:
```suggestion
  description: "Create a contact. [See docs here](https://developers.freshdesk.com/api/#create_contact)",
```

You can check and fix this in other descriptions too.

File: components/freshdesk/actions/create-contact/create-contact.mjs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  version: "0.0.2",
```
Comment (lcaresia) [inline components/freshdesk/actions/create-contact/create-contact.mjs:8]:
```suggestion
  version: "0.0.1",
```

You can check and fix this in other descriptions too.

File: components/freshdesk/actions/get-ticket/get-ticket.mjs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+      label: "Ticket ID",
+      description: "Ticket ID.",
+    },
+  },
+  async run({ $ }) {
+    const response = await this.freshdesk.getTicket({
+      $,
+      id: this.id,
+    });
+    response && $.export("$summary", "Ticket found");
```
Comment (lcaresia) [inline components/freshdesk/actions/get-ticket/get-ticket.mjs:22]:
```suggestion
    response && $.export("$summary", "Successfully found ticket");
```

File: components/freshdesk/freshdesk.app.mjs [hunk 5]
```diff
@@ -1,11 +1,232 @@
...
+
 export default {
   type: "app",
   app: "freshdesk",
-  propDefinitions: {},
+  propDefinitions: {
+    companyId: {
+      type: "integer",
+      label: "Company ID",
+      description: "ID of the primary company to which this contact belongs",
```
Comment (lcaresia) [inline components/freshdesk/freshdesk.app.mjs:14]:
```suggestion
      description: "The ID of the company",
```

File: components/freshdesk/freshdesk.app.mjs [hunk 6]
```diff
@@ -1,11 +1,232 @@
...
+      type: "integer",
+      label: "Status",
+      description: "Status of the ticket.",
+      options() {
+        return constants.TICKET_STATUS.map(({
+          label, value,
+        }) => ({
+          label,
+          value,
+        }));
```
Comment (lcaresia) [inline components/freshdesk/freshdesk.app.mjs:33]:
```suggestion
        return constants.TICKET_STATUS;
```

File: components/freshdesk/freshdesk.app.mjs [hunk 7]
```diff
@@ -1,11 +1,232 @@
...
+      type: "integer",
+      label: "Priority",
+      description: "Priority of the ticket.",
+      options() {
+        return constants.TICKET_PRIORITY.map(({
+          label, value,
+        }) => ({
+          label,
+          value,
+        }));
```
Comment (lcaresia) [inline components/freshdesk/freshdesk.app.mjs:46]:
```suggestion
        return constants.TICKET_PRIORITY;
```

File: components/freshdesk/freshdesk.app.mjs [hunk 8]
```diff
@@ -1,11 +1,232 @@
...
+    },
+    contactEmail: {
+      type: "string",
+      label: "Email",
+      description: "Contact Email.",
+      async options({ companyId }) {
+        const response = await this.getCompanies();
+        const contacts = response.filter((contact) => contact.company_id === Number(companyId));
+        return contacts.map((contact) => ({
+          label: `${contact.name} | ${contact.email}`,
```
Comment (lcaresia) [inline components/freshdesk/freshdesk.app.mjs:57]:
```suggestion
          label: contact?.email ?? contact?.name,
```

File: components/freshdesk/package.json [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  "license": "MIT",
+  "publishConfig": {
+    "access": "public"
+  },
+  "dependencies": {
+    "@pipedream/platform": "^0.9.0",
+    "async-retry": "^1.3.3",
+    "moment":"2.29.2"
+  }
+}
```
Comment (lcaresia) [inline components/freshdesk/package.json:21]:
```suggestion
}

```

File: components/freshdesk/package.json [hunk 10]
```diff
@@ -0,0 +1,21 @@
+{
+  "name": "@pipedream/freshdesk",
+  "version": "0.0.2",
```
Comment (lcaresia) [inline components/freshdesk/package.json:3]:
```suggestion
  "version": "0.0.1",
```

File: components/freshdesk/package.json [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+  "description": "Pipedream Harvest Components",
```
Comment (lcaresia) [inline components/freshdesk/package.json:4]:
```suggestion
  "description": "Pipedream Freshdesk Components",
```

File: components/freshdesk/sources/new-contact/new-contact.mjs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    for await (const contact of contacts) {
+      data.push(contact);
+    }
+    data && data.reverse().forEach((contact) => {
+      this.freshdesk.setLastDateChecked(this.db, contact.created_at);
+      if (moment(contact.created_at).isAfter(lastDateChecked)) {
+        this.$emit(contact,
+          {
+            id: contact.id,
+            summary: `Contact name: ${contact.name}`,
```
Comment (lcaresia) [inline components/freshdesk/sources/new-contact/new-contact.mjs:44]:
```suggestion
            summary: `Contact name: ${contact.name}`,
            ts: Date.parse(contact.created_at),
```

File: components/freshdesk/sources/new-ticket/new-ticket.mjs [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    for await (const ticket of tickets) {
+      data.push(ticket);
+    }
+    data && data.reverse().forEach((ticket) => {
+      this.freshdesk.setLastDateChecked(this.db, ticket.created_at);
+      if (moment(ticket.created_at).isAfter(lastDateChecked)) {
+        this.$emit(ticket,
+          {
+            id: ticket.id,
+            summary: `Ticket number: ${ticket.id}`,
```
Comment (lcaresia) [inline components/freshdesk/sources/new-ticket/new-ticket.mjs:44]:
```suggestion
            summary: `Ticket number: ${ticket.id}`,
            ts: Date.parse(ticket.created_at)
```

**Review by lcaresia — CHANGES_REQUESTED** [review]
Thanks for your contribution!

I requested just a minor changes.

File: components/freshdesk/actions/get-ticket/get-ticket.mjs [hunk 14]
```diff
@@ -0,0 +1,25 @@
+import freshdesk from "../../freshdesk.app.mjs";
+
+export default {
+  key: "freshdesk-get-ticket",
+  name: "Get Ticket Details",
+  description: "Get a Ticket. [See docs here](https://developers.freshdesk.com/api/#tickets)",
+  version: "0.0.2",
```
Comment (lcaresia) [inline components/freshdesk/actions/get-ticket/get-ticket.mjs:7]:
```suggestion
  version: "0.0.1",
```

File: components/freshdesk/sources/new-contact/new-contact.mjs [hunk 15]
```diff
@@ -0,0 +1,50 @@
+import freshdesk from "../../freshdesk.app.mjs";
+import moment from "moment";
+
+export default {
+  key: "freshdesk-new-contact",
+  name: "New Contact",
+  description: "Emit new notifications when a new contact is created",
+  version: "0.0.2",
```
Comment (lcaresia) [inline components/freshdesk/sources/new-contact/new-contact.mjs:8]:
```suggestion
  version: "0.0.1",
```

File: components/freshdesk/sources/new-ticket/new-ticket.mjs [hunk 16]
```diff
@@ -0,0 +1,50 @@
+import freshdesk from "../../freshdesk.app.mjs";
+import moment from "moment";
+
+export default {
+  key: "freshdesk-new-ticket",
+  name: "New Ticket",
+  description: "Emit new notifications when a new ticket is created",
+  version: "0.0.2",
```
Comment (lcaresia) [inline components/freshdesk/sources/new-ticket/new-ticket.mjs:8]:
```suggestion
  version: "0.0.1",
```

**Review by lcaresia — CHANGES_REQUESTED** [review]
Just need fix the versions.

**Review by lcaresia — APPROVED** [review]
LGTM!

**Comment by DilanAthukorala** [comment]
@bush4u all action requests are success but nothing gets created on freshdesk app and api responses are not returned from the action component

freshdesk url - https://pipedream-help.freshdesk.com/

```
domain - pipedream-help
apikey -  662erykZvAU7NUzbBN
```

image

**Comment by bush4u** [comment]
Kindly let me check this

**Comment by bush4u** [comment]
So, published the action, and tested.

I was able to create a company and I got API response.
I also confirmed that data was created on freshdesk portal. 

How can we narrow this down further to get the cause of the issue?



image
image

**Comment by DilanAthukorala** [comment]
@bush4u in create_contact and create_ticket actions, the `Company ID` prop is listing the contact id list. 
image


for create_ticket action params `subject , description` are **required** parameters.  
image



for create_ticket action email prop does not retrieve any emails. Should we make it an input field?
image



for create_ticket action observing following error when sending value for description_text

image



No response for list all tickets action
image

**Comment by DilanAthukorala** [comment]
Verification completed

**Pull request merged**